src/main/scala/com/azavea/modellab/Node.scala

```diff
@@ -24,60 +28,145 @@ trait Node extends Serializable {
+    buff.putInt(hashCode)
+    Base64.encodeBase64(buff.array).mkString
+  }
 }
 
-case class LoadLayer(layerName: String, layerReader: FilteringLayerReader[LayerId, SpatialKey, RasterRDD[SpatialKey]]) extends Node { 
+case class LoadLayer(layerName: String, layerReader: WindowedReader) extends Node {
   def calc(zoom: Int, bounds: GridBounds) = {
-    val rdd = layerReader.query(LayerId(layerName, zoom)).where(Intersects(bounds)).toRDD
-    rdd.foreachPartition{ p => () }
-    rdd
+    layerReader.getView(LayerId(layerName, zoom), bounds)
   }
 }
 
-case class LocalAdd(a: Node, b: Node) extends Node {
+case class FocalOp(
+  op: (Tile, Neighborhood, Option[GridBounds]) => Tile,
+  input: Node,
+  n: Neighborhood
+) extends Node {
+  require(n.extent <= 256, "Maximum Neighborhood extent is 256 cells")
+  def buffer(bounds: GridBounds, cells: Int) = GridBounds(
+    math.max(0, bounds.colMin - cells),
+    math.max(0, bounds.rowMin - cells),
+    bounds.colMax + cells,
+    bounds.rowMax + cells)
+
   def calc(zoom: Int, bounds: GridBounds) = {
-    a(zoom, bounds).combineTiles(b(zoom, bounds)) { Add(_,_) }
+    val rasterRDD = input(zoom, bounds)
+    val metaData = rasterRDD.metaData
+    val extended = buffer(bounds, 1)
+    println(s"FOCAL: $bounds -> $extended")
+    val tileRDD = FocalOperation(rasterRDD, n, Some(bounds))(op)
+      .filter { case (key, _) => bounds.contains(key.col, key.row) } // filter buffer tiles, they contain no information
+    new RasterRDD(tileRDD, metaData)
   }
 }
 
-case class LocalSubtract(a: Node, b: Node) extends Node {
+case class AspectOp(
+  op: (Tile, Neighborhood, Option[GridBounds], CellSize) => Tile,
+  input: Node,
+  n: Neighborhood
+) extends Node {
   def calc(zoom: Int, bounds: GridBounds) = {
-    a(zoom, bounds).combineTiles(b(zoom, bounds)) { Subtract(_,_) }
+    val rasterRDD = input(zoom, bounds)
+    val metaData = rasterRDD.metaData
+    val cs = metaData.layout.rasterExtent.cellSize
+    val tileRDD = rasterRDD map { case (key, tile) => key -> op(tile, n, None, cs) }
+    new RasterRDD(tileRDD, metaData)
   }
 }
 
-case class LocalMultiply(a: Node, b: Node) extends Node {
+case class SlopeOp(
+  op: (Tile, Neighborhood, Option[GridBounds], CellSize, Double) => Tile,
+  input: Node,
+  n: Neighborhood,
+  z: Double
+) extends Node {
   def calc(zoom: Int, bounds: GridBounds) = {
-    a(zoom, bounds).combineTiles(b(zoom, bounds)) { Multiply(_,_) }
+    val rasterRDD = input(zoom, bounds)
+    val metaData = rasterRDD.metaData
+    val cs = metaData.layout.rasterExtent.cellSize
+    val tileRDD = rasterRDD map { case (key, tile) => key -> op(tile, n, None, cs, z) }
+    new RasterRDD(tileRDD, metaData)
   }
 }
 
-case class LocalDivide(a: Node, b: Node) extends Node {
+case class LocalUnaryOp(
+  op: Function1[Tile, Tile],
+  input: Node
+) extends Node {
   def calc(zoom: Int, bounds: GridBounds) = {
-    a(zoom, bounds).combineTiles(b(zoom, bounds)) { Divide(_,_) }
+    val _op = op
```

Comment:
@moradology Didn't  catch this in the code review, so a heads up here. Using the `op` as a class field will cause the `LocalUnaryOp` class object to be pulled in and any of it's other fields, like the `inputs`. This will eventually spider down to `LoadLayer` class which will try to bring in `WindowReader` which is not serializable. Will end up in  Task Not Serializable exception at runtime.

Assigning the op to a local val breaks this serialization chain. In general only local variables and method arguments are safe to use. This is a property of Java serialization btw, which is used by Spark